clientv3/retry.go

```diff
@@ -23,65 +23,96 @@ import (
 	return func(rpcCtx context.Context, f rpcFunc) error {
 		for {
-			select {
-			case <-c.balancer.ConnectNotify():
-			case <-rpcCtx.Done():
-				return rpcCtx.Err()
-			case <-c.ctx.Done():
-				return c.ctx.Err()
-			}
 			pinned := c.balancer.pinned()
 			err := f(rpcCtx)
 			if err == nil {
 				return nil
 			}
 			if logger.V(4) {
-				logger.Infof("clientv3/retry: error %v on pinned endpoint %s", err, pinned)
+				logger.Infof("clientv3/retry: error %v on pinned endpoint %s (write %v)", err, pinned, isWrite)
+			}
+
+			// error from etcd server with codes.Unavailable, then endpoint switch and retry
+			// e.g. rpctypes.ErrTimeout
+			// error from etcd server with non codes.Unavailable, then no endpoint switch and no retry
+			// e.g. rpctypes.ErrEmptyKey, rpctypes.ErrNoSpace
+			rerr := rpctypes.Error(err)
+			if ev, ok := rerr.(rpctypes.EtcdError); ok {
+				if ev.Code() != codes.Unavailable {
+					return err
+				}
+				if c.balancer.endpointError(pinned, err) {
+					c.balancer.next()
+				}
+				// only retry immutable RPCs ("put at most once semantics")
+				if !isWrite {
+					select {
+					case <-c.balancer.ConnectNotify():
+					case <-rpcCtx.Done():
+						return rpcCtx.Err()
+					case <-c.ctx.Done():
+						return c.ctx.Err()
+					}
+					continue
+				}
+				return err
 			}
-			// mark this before endpoint switch is triggered
-			c.balancer.endpointError(pinned, err)
-			notify := c.balancer.ConnectNotify()
-			if s, ok := status.FromError(err); ok && s.Code() == codes.Unavailable {
-				c.balancer.next()
+
+			// if temporary gRPC connection error, then endpoint switch but no retry
+			// e.g. transport.ErrConnClosing when server closed transport, failing node
+			if tv, ok := err.(transport.ConnectionError); ok && tv.Temporary() {
+				// mark as unhealthy and wait for another endpoint to come up
+				if c.balancer.endpointError(pinned, err) {
+					c.balancer.next()
+				}
+				return err
 			}
-			if isStop(err) {
+
+			// TODO: handle transport.StreamError?
+
+			// if unknown status error from gRPC, then endpoint switch but no retry
+			s, ok := status.FromError(err)
+			if !ok {
+				if c.balancer.endpointError(pinned, err) {
+					c.balancer.next()
+				}
 				return err
 			}
-			select {
-			case <-notify:
-			case <-rpcCtx.Done():
-				return rpcCtx.Err()
-			case <-c.ctx.Done():
-				return c.ctx.Err()
+
+			// if known status error from gRPC with following codes,
+			// then endpoint switch and retry
+			if s.Code() == codes.Unavailable ||
+				s.Code() == codes.Internal ||
+				s.Code() == codes.DeadlineExceeded {
+				exist := c.balancer.endpointError(pinned, err)
+				if exist || s.Message() == "there is no address available" {
+					c.balancer.next()
+				}
+				// only retry immutable RPCs ("put at most once semantics")
+				// or mutable RPCs with "there is no address available"
+				if !isWrite || s.Message() == "there is no address available" {
+					select {
```

Comment:
we really should not duplicate this code three times in one func.

Reply:
Agree, will fix/clean-up tomorrow.